frontend/apps/reader/modules/readerbookmark.lua

```diff
@@ -197,17 +197,46 @@ end
 
 function ReaderBookmark:isBookmarkInPageOrder(a, b)
     if self.ui.document.info.has_pages then
-        if a.page == b.page then -- have bookmarks before highlights
-            return a.highlighted
+        if a.page == b.page then -- both bookmarks in the same page
+            if a.highlighted and b.highlighted then -- both are highlights, compare positions
+                local is_reflow = self.ui.document.configurable.text_wrap -- save reflow mode
+                if is_reflow == 1 then -- reflow mode doesn't set page in positions
+                    a.pos0.page = a.page
+                    a.pos1.page = a.page
+                    b.pos0.page = a.page
+                    b.pos1.page = a.page
+                end
+                self.ui.document.configurable.text_wrap = 0 -- native positions
+                -- sort start and end positions of each highlight
+                local compare_pos, a_start, a_end, b_start, b_end, result
+                compare_pos = self.ui.document:comparePositions(a.pos0, a.pos1) == 1
+                a_start = compare_pos and a.pos0 or a.pos1
+                a_end = compare_pos and a.pos1 or a.pos0
+                compare_pos = self.ui.document:comparePositions(b.pos0, b.pos1) == 1
+                b_start = compare_pos and b.pos0 or b.pos1
+                b_end = compare_pos and b.pos1 or b.pos0
+                -- compare start positions
+                compare_pos = self.ui.document:comparePositions(a_start, b_start)
+                if compare_pos == 0 then -- both highlights with the same start, compare ends
+                    result = self.ui.document:comparePositions(a_end, b_end) == -1
+                else
+                    result = compare_pos == -1
+                end
+                self.ui.document.configurable.text_wrap = is_reflow -- restore reflow mode
+                return result
+            end
+            return a.highlighted -- have page bookmarks before highlights
         end
         return a.page > b.page
     else
-        local a_page = self.ui.document:getPageFromXPointer(a.page)
-        local b_page = self.ui.document:getPageFromXPointer(b.page)
-        if a_page == b_page then -- have bookmarks before highlights
-            return a.highlighted
+        local compare_xp = self.ui.document:compareXPointers(a.page, b.page)
+        if compare_xp == 0 then -- both bookmarks with the same start
```

Comment:
> compareXPointers(a,b) being expensive

So maybe at first compare pages, and only when two bookmarks are within one page, compare xpointers.

Reply:
> I also see you didn't update the next function, `isBookmarkInReversePageOrder()`

This one is used only in one place for
https://github.com/koreader/koreader/blob/c11ea4f389bc3f64dbcaaf353f2926230a8d4916/frontend/apps/reader/modules/readerpaging.lua#L318-L320
in flipping mode, so no need of accuracy within one page.

Reply:
> This one is used only in one place for

But also by this it seems:
`skimtowidget.lua:            self:goToByEvent("GotoNextBookmarkFromPage")`

All this ReaderBookmark reorderning/insertion/binary search feels a bit crappy, thought we could get it simpler when migrating to annotations.
But my idea is that any navigation should be smooth and cheap, and only at add/remove bookmark/highlight we are allowed to do expensive stuff (expensive so it orders everything so that the others can then be smooth and cheap).

> So maybe at first compare pages, and only when two bookmarks are within one page, compare xpointers.

Yes, that would be good - if you're sure about page being ok (might not be when created in reflow mode I beleive from what you wrote in your code?)

Reply:
> any navigation should be smooth and cheap, and only at add/remove bookmark/highlight we are allowed to do expensive stuff

Let's decide. `isBookmarkInPageOrder` is used 6 times:
fixBookmarkSort
addBookmark
isBookmarkAdded
removeBookmark

getDogearBookmarkIndex
getPreviousBookmarkedPage

The first group needs accuracy. You propose to make my `isBookmarkInPageOrderFine` for it only, right?

Reply:
> The first group needs accuracy. You propose to make my isBookmarkInPositionOrder for it only, right?

Yes, exactly, that's the idea: a new function, that you would use where you needed to have stuff fixed.

- isBookmarkInPageOrder / isBookmarkInReversePageOrder for quick navigation cheap stuff where only page is enough (mainly, comparing page in binary search)
- isBookmarkInPositionOrder for fixBookmarkSort and addBookmark (which can do reordering/insertion).
- isBookmarkAdded / removeBookmark don't do any insert/ordering, so it feels they could use the cheaper one (isBookmarkAdded is used only by DictQuickLookup).

When you first talked about this PR some days ago, I though you might need a one step migration: fixBookmarkSort() did a migration (and may do it for books not opened since long), so you might want to either make it better and have it be called again (if config:hasNot("bookmarks_sorted2")), or have another fixBookmarkSort2() so your isBookmarkInPositionOrder() can work on something already perfectly sorted (or may be it doesn't need it ? and will do the sorting? but then, not until another bookmark is added?).

Reply:
Okay. What about `removeBookmark`? It makes the full scan when there are several bookmarks in a page.
Using `isBookmarkInPositionOrder` we can remove full scan code, it is not used (I've checked).
https://github.com/koreader/koreader/blob/ec280f874f0cfce660bdb4d396589939503d3c8b/frontend/apps/reader/modules/readerbookmark.lua#L903-L906

Reply:
That means you would need to have your bookmarks correctly sorted with your new way, after an initial migration ?
But for removeBookmark(), I wonder if doing binary search and calling your (expensive) compareXPointers() might be slower than a full scan :) (but may be that's no more an issue as you first check the are identical before using it).

(I think it's me that added the fallback full scan, may be because I was in a situation that shouldn't happen once you have precisely sorted the bookmark ? Dunno.)